test/test_endpoint_inferral.cpp

```diff
@@ -0,0 +1,73 @@
+#include <datadog/endpoint_inferral.h>
+
+#include <string>
+
+#include "test.h"
+
+using namespace datadog::tracing;
+
+#define TEST_ENDPOINT(x) TEST_CASE(x, "[endpoint_inferral]")
+
+TEST_ENDPOINT("invalid inputs and root") {
+  CHECK(infer_endpoint("") == "/");
+  CHECK(infer_endpoint("abc") == "/");
+  CHECK(infer_endpoint("/") == "/");
+  CHECK(infer_endpoint("////") == "/");
+}
+
+TEST_ENDPOINT("skips empty components") {
+  CHECK(infer_endpoint("/a//b") == "/a/b");
+  CHECK(infer_endpoint("/a/b/") == "/a/b/");
+}
+
+TEST_ENDPOINT("int and int_id replacement") {
+  CHECK(infer_endpoint("/users/12") == "/users/{param:int}");
+  CHECK(infer_endpoint("/v1/0-1_2.3") == "/v1/{param:int_id}");
+  CHECK(infer_endpoint("/x/09") == "/x/09");  // leading zero not int
+  CHECK(infer_endpoint("/1") == "/1");        // single digit not int/int_id
+}
+
+TEST_ENDPOINT("hex and hex_id replacement") {
+  CHECK(infer_endpoint("/x/abcde9") == "/x/{param:hex}");
+  CHECK(infer_endpoint("/x/ab_cd-9") == "/x/{param:hex_id}");
+}
+
+TEST_ENDPOINT("str replacement by special or length") {
+  CHECK(infer_endpoint("/x/a%z") == "/x/{param:str}");
+  std::string longseg(20, 'a');
+  std::string path = std::string("/x/") + longseg;
+  CHECK(infer_endpoint(path) == "/x/{param:str}");
+}
+
+TEST_ENDPOINT("other specials yield str") {
```

Comment:
small nit: Why not use this test for all special characters (including the % character) and make this test only special characters and the previous one only based on string length?